String.cxx

```diff
@@ -0,0 +1,184 @@
+/*
+ *  String.cxx
+ *  Copyright 2020 ItJustWorksTM
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <algorithm>
+#include <cctype>
+#include <charconv>
+#include <cstring>
+#include <iostream>
+#include <regex>
+#include <string>
+
+enum StringBaseConv {
+    BIN = 2,
+    DEC = 10,
+    HEX = 16,
+};
+
+enum class byte : unsigned char {};
+
+class String {
+    std::string m_u;
+
+    String(std::string u) : m_u{std::move(u)} {}
+
+  public:
+    String() noexcept = default;
+    String(const String&) = default;
+    String(String&&) noexcept = default;
+    String& operator=(const String&) = default;
+    String& operator=(String&&) = default;
+
+    template <std::size_t N> inline String(const char (&charr)[N]) : m_u{charr, N} {}
+    inline String(const char* cstr) : m_u{cstr} {}
+    inline explicit String(char c) : m_u{1, c} {}
+    template <class T, class = std::enable_if_t<std::is_integral<T>::value>> inline String(T val, StringBaseConv base) {
+        constexpr static auto base2len_upperbound = [](StringBaseConv base) /*constexpr*/ {
+            switch (base) {
+            case BIN:
+                return 65;
+            case DEC:
+
+            case HEX:
+                return 17;
+            default:
+                throw;
+            }
+        };
+        m_u.resize(base2len_upperbound(base));
+        const auto res = std::to_chars(&*m_u.begin(), &*m_u.rbegin(), val, +base);
```

Comment:
std::to_chars(m_u.data(), m_u.data() + m_u.size(), val, +base); 
should be more sane I reckon.

Reply:
And what does the + do in +base ?

Reply:
`+base` is integral promotion

Reply:
How is something with an sum right the middle more sane?

Reply:
Its more readable imo, Its the same but actually shows better what it is compared to &*m_u.rbegin().
I don't mind either but prefer things more written out if its not too long.

Reply:
My problem with the pointer arithmetic is that it's a huge source of errors